[wip]Database backed stickybans

MrsTonedOne opened the pull request:
If legacy bans is not active, stickybans are stored in the database and applied at initialization time.

Also Supports:
* Disabling stickybans for a round. (timeout)
* Exempting a key from matching a stickyban, 
* Detecting rogue stickybans before anybody even gets disconnected. (new matches trigger a 5 second sleep and abort enforcement if enough other new matches happen in that timeframe)

Todo: testing. (gotta beat the freeze!)

fixes #12068

MrsTonedOne committed:
Database backed stickybans

Supports disabling stickybans for a round, exempting a key from matching a stickyban, and it now also detects rogue stickybans before anybody currently connected even gets disconnected. (new matches trigger a 5 second sleep and abort enforcement if enough other new matches happen in that timeframe)

Jordie0608 commented on code/modules/admin/IsBanned.dm:
```diff
@@ -142,40 +164,66 @@
 
+			if (ban["fromdb"])
+				if(!CONFIG_GET(flag/ban_legacy_system) && SSdbcore.Connect())
+					var/datum/DBQuery/query_add_match = SSdbcore.NewQuery("INSERT IGNORE INTO [format_table_name("stickyban_matched_ckey")] (matchedckey, stickyban) VALUES ('[sanitizeSQL(ckey)]', '[sanitizeSQL(bannedckey)]')")
```
Also why `IGNORE`, won't an empty field cause errors elsewhere?

MrStonedOne replied:
ignore also ignores duplicate keys instead just stopping the query.

Jordie0608 commented on code/modules/admin/stickyban.dm:
```diff
@@ -124,14 +130,170 @@
+			SSstickyban.cache[ckey] = ban
+
+
+			var/datum/DBQuery/query_exempt_stickyban_alt = SSdbcore.NewQuery("UPDATE [format_table_name("stickyban_matched_ckey")] SET exempt = 1 WHERE stickyban = ckey = '[sanitizeSQL(ckey)]' AND matched_ckey = '[sanitizeSQL(alt)]'")
```
`WHERE stickyban = ckey = '[sanitizeSQL(ckey)]'`

Jordie0608 commented on code/controllers/subsystem/stickyban.dm:
```diff
@@ -1,34 +1,82 @@
+	var/datum/DBQuery/query_stickyban_matches = SSdbcore.NewQuery("SELECT stickyban, matched_ckey, first_matched, exempt FROM [format_table_name("stickyban_matrched_ckey")]")
+	if (!query_stickyban_matches.warn_execute())
+		return
+	query_stickyban_matches.SetConversion(4, SSdbcore.NUMBER_CONV) //read exempt as a number, not a string
```
I tried to figure out what this proc is supposed to do but instead it just runtimed because 3 isn't a list.

optimumtact commented on code/modules/admin/IsBanned.dm:
```diff
@@ -126,10 +126,32 @@
+					return
+				bancache[ban["ckey"]] = world.GetConfig("ban", ban["ckey"])
+				//Hacky way to ensure somebody exempt from one stickyban doesn't get exempt from all stickybans
+				world.SetConfig("ban", bannedckey, null)
```
how does this work?

MrStonedOne replied:
We remove the current ban, then re-call parent to see if it matches any other stickybans.

At the end we re-add the ban(s) that were removed.

We cache the checks because byond will re-call isbanned() when we re-add the stickybans. I need to add some checks against that happening so the rest of isbanned() doesn't get called.

MrsTonedOne committed:
Some fixes.

Jordie0608 commented:
Does this address #12068 (if that's even still an issue)?

MrsTonedOne committed:
Adds automatic importing of older legacy stickybans

MrStonedOne commented on code/modules/admin/stickyban.dm:
```diff
@@ -150,14 +312,21 @@
+/datum/admins/proc/stickyban_gethtml(ckey)
+	var/ban = get_stickyban_from_ckey(ckey)
+	if (!ban)
+		return
```
yes @Jordie0608 this fixes that

optimumtact commented:
what is this waiting for

MrStonedOne commented:
I forgot actually, but i have some free time today so i'll poke it and hope i figure out where the final bug was.

MrStonedOne commented:
You can too predictably make stickybans go rogue to allow such a system to exist. Hopfully lummox fixes it soon, but he said he wouldn't be able to look into it until next week.

MrStonedOne commented:
I need to finish patching up the bug and push it up then this will be ready

MrsTonedOne committed:
Fixes and improvements. Hub password changes

Uses lazy access on inner lists (had to make lazy access work on inner lists @cyberboss you won't like this but lazyaccess is useless if you can't use it on inner lists)
Byond only does full stickyban matching when the server has a valid hub password set or compiled in, so now we have to remove it during init.

KorPhaeron commented:
reopen this when you're ready

MrStonedOne commented:
its got a commit from 3 days ago don't believe github's timestamps

KorPhaeron commented:
thanks github

optimumtact commented:
@MrStonedOne come on man

optimumtact commented:
@MrStonedOne come on man

Cyberboss commented:
>reopening someone's half year old stale pr
